tarantool/dbapi.py

```diff
@@ -0,0 +1,227 @@
+        for params in param_sets:
+            self.execute(query, params)
+            if self.rowcount == -1:
+                rowcount = -1
+            if rowcount != -1:
+                rowcount += self.rowcount
+        self._rowcount = rowcount
+
+    @property
+    def lastrowid(self):
+        """
+        This read-only attribute provides the rowid of the last modified row
+        (most databases return a rowid only when a single INSERT operation is
+        performed).
+        """
+        return self._lastrowid
+
+    @property
+    def rowcount(self):
+        """
+        This read-only attribute specifies the number of rows that the last
+        .execute*() produced (for DQL statements like SELECT) or affected (
+        for DML statements like UPDATE or INSERT).
+        """
+        return self._rowcount
+
+    def _check_result_set(self, error=None):
+        """
+        Closed method for raising an error when Cursor object does not have
+        any row to fetch. Useful for checking access after DQL requests.
+        """
+        if self._rows is None:
+            raise ProgrammingError(error or "Result set is empty.")
+
+    def fetchone(self):
+        """
+        Fetch the next row of a query result set, returning a single
+        sequence, or None when no more data is available.
+        """
+        self._check_result_set()
+        return self.fetchmany(1)[0] if self._rows else None
+
+    def fetchmany(self, size=None):
+        """
+        Fetch the next set of rows of a query result, returning a sequence of
+        sequences (e.g. a list of tuples). An empty sequence is returned when
+        no more rows are available.
+        """
+        self._check_result_set()
+
+        size = size or self.arraysize
+
+        if len(self._rows) < size:
+            items = self._rows
+            self._rows = []
+        else:
+            items, self._rows = self._rows[:size], self._rows[size:]
+
+        return items
+
+    def fetchall(self):
+        """Fetch all (remaining) rows of a query result, returning them as a
+        sequence of sequences (e.g. a list of tuples). Note that the cursor's
+        arraysize attribute can affect the performance of this operation.
+        """
+        self._check_result_set()
+
+        items = self._rows
+        self._rows = []
+        return items
+
+    def setinputsizes(self, sizes):
+        """PEP-249 allows to not implement this method and do nothing."""
+
+    def setoutputsize(self, size, column=None):
+        """PEP-249 allows to not implement this method and do nothing."""
+
+
+class Connection(BaseConnection):
+
+    @property
+    def server_version(self):
+        """
+        Returns the server version from greeting
+        """
+        return getattr(self, 'version_id')
+
+    def _check_not_closed(self, error=None):
+        """
+        Checks if the connection is not closed and rises an error if it is.
+        """
+        if self._socket is None:
```

Comment:
I would use is_closed() method here, because all code that manages self._socket is in the base class. Otherwise it may be hard to keep consistency if something will be refactored around self._socket.

Reply:
Changed with .is_closed() method. Thanks for the clarification.